numba/tests/test_entrypoints.py

```diff
@@ -31,45 +37,38 @@ def test_init_entrypoint(self):
         # loosely based on Pandas test from:
         #   https://github.com/pandas-dev/pandas/pull/27488
 
-        # FIXME: Python 2 workaround because nonlocal doesn't exist
-        counters = {'init': 0}
-
-        def init_function():
-            counters['init'] += 1
-
-        mod = types.ModuleType("_test_numba_extension")
-        mod.init_func = init_function
+        mod = mock.Mock(__name__='_test_numba_extension')
 
         try:
             # will remove this module at the end of the test
             sys.modules[mod.__name__] = mod
 
             # We are registering an entry point using the "numba" package
-            # ("distribution" in pkg_resources-speak) itself, though these are
+            # ("distribution" in importlib-speak) itself, though these are
             # normally registered by other packages.
-            dist = "numba"
-            entrypoints = pkg_resources.get_entry_map(dist)
-            my_entrypoint = pkg_resources.EntryPoint(
-                "init",  # name of entry point
-                mod.__name__,  # module with entry point object
-                attrs=['init_func'],  # name of entry point object
-                dist=pkg_resources.get_distribution(dist)
+            my_entrypoint = importlib_metadata.EntryPoint(
+                'init', '_test_numba_extension:init_func', 'numba_extensions',
             )
-            entrypoints.setdefault('numba_extensions',
-                                   {})['init'] = my_entrypoint
 
-            from numba.core import entrypoints
-            # Allow reinitialization
-            entrypoints._already_initialized = False
+            with mock.patch.object(
+                importlib_metadata,
+                'entry_points',
+                return_value={'numba_extensions': (my_entrypoint,)},
+            ):
+
+                from numba.core import entrypoints
 
-            entrypoints.init_all()
+                # Allow reinitialization
+                entrypoints._already_initialized = False
```

Comment:
Whilst not new, this mutates module global state, it should probably run in a `try: finally:` with the `finally:` setting the state back to whatever it was before the `try`.

Reply:
Is that really the right thing to do? If we then call `entrypoints.init_all()` then the entry points will have been initialized, so it feels like it makes more sense to leave `_already_initialized` as whatever `init_all()` leaves it as.

It seems that generally this test assumes the absence of other entry points for Numba extensions - ensuring we don't actually re-initialize an unrelated extension would seem to need us to keep track of which extensions were already initialized in `numba.core.entrypoints`, which is a bit more than the scope of this PR so far.

Reply:
Thanks for expanding on this, I agree, the right thing here is probably to track opposed to add things to mitigate. This is indeed out of scope for this PR.